sections/sprint-3/goal-position-generation.tex

```diff
@@ -0,0 +1,25 @@
+\section{Goal generation}\label{subsec:goalrefactoring}
+The first implementation of goal generation was done entirely in Unity.
+This means that the goal generation and detection would have been handled on the client-side.
+The idea was to limit the amount of data that needed to be sent across the network.
+But when it was decided that new goal zones should be generated in random locations after a goal had been scored, generating it on the client-side could cause the game state to be different between different players.
+Either by some players not receiving all the necessary ball positions to observe a goal or by generating goal zones in different places.
+One way to solve this was by giving all clients the same seed at the start of the match to generate random positions based of so that the goal zones would be in the same positions across all clients.
+The concern about this solution was that the goal detection could still have issues if some of the positional data for the ball would not arrive at the same time or at all.
+In the end, it was decided to move all the goal zone generation and goal detection to the server-side and utilizing TCP so that the data would be consistent across all clients.
+
+\subsection{Server side}
+On the server-side, the size of the goal is calculated based on a percentage of the shortest edge of the playing field.
+This size is used to calculate where the center points of the goal zones can be placed as the corners of the goal zones should not exceed the playing field edges.
+The goal zone will always be a square where all edge lengths are of equal length and the playing field should always closely resemble a rectangle.
+The first goal zones are always set at a certain point on the playing field, afterward, the goal zones positions are randomly generated.
+The randomly generated goal zones will always mirror each other and will not spawn to close to the middle of the playing field.
+The goal zones will also never change sides so if a goal zone is initially generated on the left of the center of the playing field it will stay on that side for the duration of the game.
+Whenever a goal is scored new center points are generated for both goal zones and then send from the server to all the clients connected.
+
+\subsection{Client side}
+On the client-side, the new center points are received along with the goal size.
+Based on this data the corners of the goal zones can be calculated.
+The corners get calculated in clockwise order, this is done to make it easier to apply the mesh which is the visual component in unity that renders the goal zone so that the players can actually see it.
```

Comment:
```suggestion
The corners are calculated in clockwise order.
This is done to make it easier to apply the mesh which is the visual component in unity that renders the goal zone so that the players can actually see it.
```